ballerina/common_utils.bal

```diff
@@ -351,3 +352,71 @@ isolated function getInputObjectFieldFormPath((string|int)[] path, string name)
+    if clientError is http:ApplicationResponseError {
+        anydata|error data = clientError.detail().get("body").ensureType(anydata);
+        if data is error {
+            return error RequestError("GraphQL Client Error", data);
+        }
+        return error RequestError("GraphQL Client Error", body = data);
+    }
+    return error RequestError("GraphQL Client Error", clientError);
+}
+
+isolated function handleGraphqlErrorResponse(map<json> responseMap) returns RequestError|ServerError {
+    ErrorDetail[]|error errors = responseMap.get("errors").cloneWithType(GraphQLErrorArray);
+    if errors is error {
+        return error RequestError("GraphQL Client Error", errors);
+    }
+    json? data = (responseMap.hasKey("data")) ? responseMap.get("data") : ();
+    map<json>? extensions = (responseMap.hasKey("extensions")) ? (responseMap.get("extensions") is () ? () : 
+        <map<json>> responseMap.get("extensions")) : ();
+    return error ServerError("GraphQL Server Error", errors = errors, data = data, extensions = extensions);
+}
+
+isolated function performDataBinding(typedesc<GenericResponse|record{}|json> targetType, json graphqlResponse) 
+                                     returns GenericResponse|record{}|json|RequestError {
+    do {
+        if targetType is typedesc<GenericResponse> {
+            GenericResponse response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        } else if targetType is typedesc<record{}> {
+            record{} response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        } else if targetType is typedesc<json> {
+            json response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        }
+    } on fail error e {
+        return error RequestError("GraphQL Client Error",  e);
+    }
+    return error RequestError("GraphQL Client Error, Invalid binding type.");
+}
+
+isolated function performDataBindingWithErrors(typedesc<GenericResponseWithErrors|record{}|json> targetType, 
+                                               json graphqlResponse) 
+                                               returns GenericResponseWithErrors|record{}|json|RequestError {
+    do {
+        if targetType is typedesc<GenericResponseWithErrors> {
+            GenericResponseWithErrors response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        } else if targetType is typedesc<record{}> {
+            record{} response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        } else if targetType is typedesc<json> {
+            json response = check graphqlResponse.cloneWithType(targetType);
+            return response;
+        }
+    } on fail error e {
+        return error RequestError("GraphQL Client Error",  e);
+    }
+    return error RequestError("GraphQL Client Error, Invalid binding type.");
```

Comment:
DItto

Reply:
Applied the suggested logic. Gives the following error.

```
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2022-03-31 09:04:51,391] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app')
        at org.wso2.ballerinalang.compiler.util.Unifier.getConstraintTypeIfNotError(Unifier.java:670)
        at org.wso2.ballerinalang.compiler.util.Unifier.getTypeAddingArgIfNotProvided(Unifier.java:648)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:572)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:77)
        at org.wso2.ballerinalang.compiler.semantics.model.types.BParameterizedType.accept(BParameterizedType.java:65)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:439)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:77)
        at org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType.accept(BUnionType.java:143)
        at org.wso2.ballerinalang.compiler.util.Unifier.build(Unifier.java:99)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationArgs(TypeChecker.java:6728)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParam(TypeChecker.java:6481)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParamAndReturnType(TypeChecker.java:6360)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.getLangLibMethod(TypeChecker.java:6353)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInLangLib(TypeChecker.java:3264)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visitInvocation(TypeChecker.java:2970)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2944)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:126)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:380)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:354)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:4014)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangReturn.accept(BLangReturn.java:48)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1902)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:2504)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangIf.accept(BLangIf.java:60)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1902)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:3691)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangDo.accept(BLangDo.java:69)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:524)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:500)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:76)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4020)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:349)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:309)
        at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:188)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:436)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:386)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:213)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:198)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:114)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:109)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:233)
        at io.ballerina.projects.Package.getCompilation(Package.java:145)
        at io.ballerina.projects.Package.runCodeGeneratorPlugins(Package.java:250)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:84)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.TestCommand.execute(TestCommand.java:265)
        at java.base/java.util.Optional.ifPresent(Unknown Source)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [graphql:(1:1,1:1)] Compilation failed due to: class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app')
error: compilation contains errors
```